### PR TITLE
Feature/dgj 416 deleting completed submissions

### PR DESCRIPTION
**Pull request by iman-jamali-fw**
## Summary

PR to delete completed/in-completed submissions and applications. Changes were made based on #416 and #417 

## Changes
- An endpoint was added to WebApi to delete the applications
- The form.io delete function was changed to support physical delete (previously soft delete)
- A listener was added to Camunda to delete the submission in the workflow
- A listener was added to Camunda to delete the application in the workflow

## Screenshots (if applicable)
Screen Shot 2022-07-26 at 7 46 58 AM

## Notes
Please note that `DeleteApplicationListener` has an issue that prevents the call to delete the application from completing. This will be fixed in the future PRs 🤞🏼

### CODE REVIEW

File: forms-flow-forms/src/util/delete.js [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+   *   The submission id to flag as deleted.
+   * @param {Array} forms
+   *   A list of form ids to flag all submissions as deleted.
+   * @param {Object} req
+   *   The express request object.
+   *
+   * @returns {Promise}
+   *   Result of deleting the submission(s), resolved by deleted entity(ies).
+   */
+  function deleteSubmission(subId, forms, req) {
```
Comment (iman-jamali-fw) [inline forms-flow-forms/src/util/delete.js:25]: This function was changed to support physical delete

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java [hunk 2]
```diff
@@ -174,7 +186,10 @@ public Map<String,Object> retrieveFormValues(String formUrl, boolean withFileInf
             while (dataElements.hasNext()) {
                 Map.Entry<String, JsonNode> entry = dataElements.next();
 
-                if(!withFileInfo && entry.getValue() != null && entry.getValue().isArray() && entry.getValue().get(0).has("originalName")) {
+                if(!withFileInfo && entry.getValue() != null 
+                    && entry.getValue().isArray() 
+                    && entry.getValue().get(0) != null 
+                    && entry.getValue().get(0).has("originalName")) {
```
Comment (iman-jamali-fw) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:192]: This prevents an existing issue with a null point exception.

**Review by david-sabadash — APPROVED** [review]

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteSubmissionListener.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public void notify(DelegateTask delegateTask) {
+        try {
+            deleteSubmission(delegateTask.getExecution());
+        } catch (IOException e) {
+            handleException(delegateTask.getExecution(), ExceptionSource.TASK, e);
+        }
+    }
+
+    private void deleteSubmission(DelegateExecution execution) throws IOException {
+        System.out.println("DeleteSubmissionListener.deleteSubmission Started");
```
Comment (simensma-fresh) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteSubmissionListener.java:36]: Sysout

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java [hunk 4]
```diff
@@ -81,6 +81,18 @@ public String createSubmission(String formUrl, String submission) throws IOExcep
         }
     }
 
+    public void deleteSubmission(String submissionUrl) throws IOException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        ResponseEntity<String> response = httpServiceInvoker.execute(submissionUrl, HttpMethod.DELETE, null);
+        if (response.getStatusCode().value() == HttpStatus.OK.value()) {
+            System.out.println("Submission was deleted successfully: " +  submissionUrl);
```
Comment (simensma-fresh) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:88]: Sysout

**Review by simensma-fresh — CHANGES_REQUESTED** [review]

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteApplicationListener.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            handleException(delegateTask.getExecution(), ExceptionSource.TASK, e);
+        }
+    }
+
+    private String getApplicationUrl(DelegateExecution execution){
+        return httpServiceInvoker.getProperties().getProperty("api.url")+"/application/"+execution.getVariable("applicationId");
+    }
+
+    private void deleteApplication(DelegateExecution execution) throws IOException {
+        System.out.println("DeleteApplication started");
```
[the remark this answers is not in the document]
Comment (iman-jamali-fw) [reply forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteApplicationListener.java:51]: I added logs for debugging and testing when it is deployed to DEV, which will delete after a successful test on DEV.

[the remark this answers is not in the document]
Comment (simensma-fresh) [reply forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteApplicationListener.java:51]: I think they will likely not show up in dev/test due to the debug levels set, so you might have to switch this to using a logger instead.

```java
    private final Logger LOGGER = LoggerFactory.getLogger(DeleteapplicationListener.class);

   LOGGER.info("blah");
```

File: forms-flow-api/src/formsflow_api/resources/application.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+                "message": "Invalid request data",
+            }, HTTPStatus.BAD_REQUEST
+
+            current_app.logger.warning(response)
+            current_app.logger.warning(submission_err)
+
+            return response, status
+
+    @staticmethod
+    @auth.require
```
[the remark this answers is not in the document]
Comment (iman-jamali-fw) [reply forms-flow-api/src/formsflow_api/resources/application.py:182]: Addressed

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteApplicationListener.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
+package org.camunda.bpm.extension.hooks.listeners;
```
[the remark this answers is not in the document]
Comment (iman-jamali-fw) [reply forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/DeleteApplicationListener.java:1]: Sounds good. Addressed

**Review by simensma-fresh — APPROVED** [review]

**Pull request merged**